Objects/longobject.c

```diff
@@ -15,35 +15,78 @@
 #include <float.h>
 #include <stddef.h>
 #include <stdlib.h>               // abs()
+#include <stdbool.h>
+#include <string.h>
+#include <limits.h>
 
 #include "clinic/longobject.c.h"
 /*[clinic input]
 class int "PyObject *" "&PyLong_Type"
 [clinic start generated code]*/
 /*[clinic end generated code: output=da39a3ee5e6b4b0d input=ec0275e3422a36e3]*/
 
-/* Is this PyLong of size 1, 0 or -1? */
-#define IS_MEDIUM_VALUE(x) (((size_t)Py_SIZE(x)) + 1U < 3U)
+#define IS_SMALL_INT(ival) (-_PY_NSMALLNEGINTS <= (ival) && (ival) < _PY_NSMALLPOSINTS)
+#define IS_SMALL_UINT(ival) ((ival) < _PY_NSMALLPOSINTS)
 
-/* convert a PyLong of size 1, 0 or -1 to a C integer */
-static inline stwodigits
-medium_value(PyLongObject *x)
+/* FIXME(lpereira): These predicates/utility functions can be marked with __attribute__((pure)) so that,
+ * when inlined, part of their computation can be avoided if called repeatedly with the same argument. */
+Py_LOCAL_INLINE(bool)
+fits_in_size_field(Py_ssize_t size)
 {
-    assert(IS_MEDIUM_VALUE(x));
-    return ((stwodigits)Py_SIZE(x)) * x->ob_digit[0];
+    /* 1 bit for the sign bit, 1 bit for the "is inlined or bignum" flag */
+    return (Py_size_t)size < 1UL<<(sizeof(size) * CHAR_BIT - 2);
 }
 
-#define IS_SMALL_INT(ival) (-_PY_NSMALLNEGINTS <= (ival) && (ival) < _PY_NSMALLPOSINTS)
-#define IS_SMALL_UINT(ival) ((ival) < _PY_NSMALLPOSINTS)
+Py_LOCAL_INLINE(bool)
+as_inlined_int(const PyLongObject *x, Py_ssize_t *value)
+{
+    *value = (Py_ssize_t)((Py_size_t)x->ob_size >> 1UL);
```

Comment:
To what extent do we care about the implementation-definedness of shifting negative integers?

As in https://wiki.sei.cmu.edu/confluence/plugins/servlet/mobile?contentId=87152418#content/view/87152374

Reply:
Check out `Py_ARITHMETIC_RIGHT_SHIFT`.

Reply:
`ob_size` is cast to an unsigned quantity (`Py_size_t` is unsigned), then shifted, then cast back to a signed quantity of the same size (`Py_ssize_t`).  This is defined.